core/src/main/java/io/grpc/Metadata.java

```diff
@@ -254,6 +254,7 @@ private String toStringInternal() {
   public static class Headers extends Metadata {
     private String path;
     private String authority;
+    private GrpcSession session;
```

Comment:
I'm not quite settled on whether to use Headers or not for storing the session. The primary alternative would be ServerCall.

I had been hoping to get rid of Headers/Trailers and just have Metadata, which would be made possible by https://github.com/grpc/grpc-java/issues/150 . The core issue is that Headers needs to store different things for clients -> server vs server -> clients, and client vs server. For example, ```session``` here would only be used for server-side on client -> server; it doesn't make sense for the other three cases.